presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergFilterPushdown.java

```diff
@@ -0,0 +1,219 @@
+import static com.facebook.presto.spi.ConnectorPlanRewriter.rewriteWith;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Predicates.not;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergFilterPushdown
+        implements ConnectorPlanOptimizer
+{
+    private final RowExpressionService rowExpressionService;
+    private final StandardFunctionResolution functionResolution;
+    private final FunctionMetadataManager functionMetadataManager;
+    protected final IcebergResourceFactory resourceFactory;
+    protected final HdfsEnvironment hdfsEnvironment;
+    protected final TypeManager typeManager;
+    protected IcebergTransactionManager icebergTransactionManager;
+
+    public IcebergFilterPushdown(
+            RowExpressionService rowExpressionService,
+            StandardFunctionResolution functionResolution,
+            FunctionMetadataManager functionMetadataManager,
+            IcebergTransactionManager transactionManager,
+            IcebergResourceFactory resourceFactory,
+            HdfsEnvironment hdfsEnvironment,
+            TypeManager typeManager)
+    {
+        this.rowExpressionService = requireNonNull(rowExpressionService, "rowExpressionService is null");
+        this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
+        this.functionMetadataManager = requireNonNull(functionMetadataManager, "functionMetadataManager is null");
+        this.icebergTransactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.resourceFactory = requireNonNull(resourceFactory, "resourceFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode maxSubplan, ConnectorSession session, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        if (!isPushdownFilterEnabled(session)) {
+            return maxSubplan;
+        }
+        return rewriteWith(new SubfieldExtractionRewriter(
+                session,
+                idAllocator,
+                rowExpressionService,
+                functionResolution,
+                functionMetadataManager,
+                icebergTransactionManager,
+                resourceFactory,
+                hdfsEnvironment,
+                typeManager), maxSubplan);
+    }
+
+    public static class SubfieldExtractionRewriter
+            extends BaseSubfieldExtractionRewriter
+    {
+        protected final IcebergResourceFactory resourceFactory;
+        protected final HdfsEnvironment hdfsEnvironment;
+        protected final TypeManager typeManager;
+        protected IcebergTransactionManager icebergTransactionManager;
```

Comment:
Ditto. icebergTransactionManager doesn't need to be persisted as a class member.

Reply:
I agree, I have removed the class member here.